### PR TITLE
!!! TASK: Refactor `Node` property mapper to use new NodeAddress

### PR DESCRIPTION
**Pull request by mhsdesign**
Solves: https://github.com/neos/neos-development-collection/issues/4873

This is an intermediate step after [!!! FEATURE: Overhaul node uri building](https://github.com/neos/neos-development-collection/pull/4892) before we fully move to the new node address serialisation in the Neos Ui.
Everything was refactored here that does not require adjustments in the Neos Ui (and the workspace module was also not really touched).

- avoid using the legacy node address instance in the code. The `NodeAddressFactory` is only used at the places where we still depend on the legacy format, but we create already new node address instances via `createCoreNodeAddressFromLegacyUriString`
- removed the $contentStreamId field from the legacy NodeAddress already (as we can otherwise not transform a new node address to this one). Also this prevents the $contentStreamId field from accidentally being used again in the migration time.
- removed the the unused legacy node address `isInLiveWorkspace` method.
- The currently unused method `Neos.Node.serializedNodeAddress(node)` will return the new node address format
- where previously `Node $node` was used to trigger property mapping (`DataSourceController` and `WorkspaceController`) we now pass the node as `string` and use the legacy node address factory to handle the legacy format manually. A followup will adjust everything to work with the new node address.
- the unused linkinservice will use the new node address string format if a string is passed as `$node`
- the Node property mappers were adjusted to use the new node address and not depend on the global active request handler
- lint `Neos.Workspace.Ui/Classes` with phpstan

**Upgrade instructions**

This change is breaking for existing Neos 9 users as the node being property mapped will require a new serialised input.

```php
public function someAction(Node $node): {}
```

In Neos 8.3 you could call this action with the node like

```
?node=/sites/neosdemo/features/multiple-columns@user-admin;language=en_US
```

in Neos9-dev it was

```
?node=user-admin__eyJsYW5ndWFnZSI6ImVuX1VTIn0%3D__995c9174-ddd6-4d5c-cfc0-1ffc82184677
```

and now it must be the node address in  json `NodeAddress::toJson()`

```
?node=%7B%22contentRepositoryId%22%3A%22default%22%2C%22workspaceName%22%3A%22live%22%2C%22dimensionSpacePoint%22%3A%7B%22language%22%3A%22es%22%2C%22country%22%3A%22ar%22%7D%2C%22aggregateId%22%3A%22node-1%22%7D
```

to build such a uri you can simply use:

```php
$myActionUri = new Uri($this->uriBuilder->uriFor(...));
$myActionUriWithNode = UriHelper::uriWithAdditionalQueryParameters(
    $myActionUri,
    ['node' => NodeAddress::fromNode($node)->toJson()]
);
```

**Review instructions**

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

File: Neos.Neos/Classes/Controller/Backend/ContentController.php [hunk 1]
```diff
@@ -150,14 +150,15 @@ public function uploadAssetAction(Asset $asset, string $metadata, string $node,
         $contentRepositoryId = SiteDetectionResult::fromRequest($this->request->getHttpRequest())
             ->contentRepositoryId;
         $contentRepository = $this->contentRepositoryRegistry->get($contentRepositoryId);
-        $nodeAddress = NodeAddressFactory::create($contentRepository)->createFromUriString($nodeAddressString);
+        // todo legacy uri node address notation used. Should be refactored to use json encoded NodeAddress
```
Comment (bwaidelich) [inline Neos.Neos/Classes/Controller/Backend/ContentController.php:153]: The TODO is related to line 149, right?

File: Neos.Neos/Classes/FrontendRouting/NodeAddress.php [hunk 2]
```diff
@@ -22,16 +22,6 @@
...
- * A persistent, external "address" of a node; used to link to it.
- *
- * Describes the intention of the user making the current request:
- * Show me
- *  node $nodeAggregateId
- *  in dimensions $dimensionSpacePoint
- *  in contentStreamId $contentStreamId
- *
- * It is used in Neos Routing to build a URI to a node.
- *
```
Comment (bwaidelich) [inline Neos.Neos/Classes/FrontendRouting/NodeAddress.php:34]: Why did you remove those?

File: Neos.Neos/Classes/FrontendRouting/NodeAddress.php [hunk 3]
```diff
@@ -42,8 +32,9 @@
     /**
      * @internal use NodeAddressFactory, if you want to create a NodeAddress
      */
+    /** @phpstan-ignore-next-line its all just temporary */
```
Comment (bwaidelich) [inline Neos.Neos/Classes/FrontendRouting/NodeAddress.php:35]: Could you elaborate the comment a bit? Why do we need it?

File: Neos.Neos/Classes/FrontendRouting/NodeAddress.php [hunk 4]
```diff
@@ -42,8 +32,9 @@
...
     public function __construct(
-        public ContentStreamId $contentStreamId,
+        ?ContentStreamId $_contentStreamId,
```
Comment (bwaidelich) [inline Neos.Neos/Classes/FrontendRouting/NodeAddress.php:37]: I assume, you just kept this to keep the PR smaller? But is it necessary to rename the parameter?

File: Neos.Neos/Classes/FrontendRouting/NodeAddress.php [hunk 5]
```diff
@@ -59,16 +50,10 @@ public function serializeForUri(): string
...
-    {
-        return $this->workspaceName->isLive();
-    }
-
     public function __toString(): string
     {
         return sprintf(
-            'NodeAddress[contentStream=%s, dimensionSpacePoint=%s, nodeAggregateId=%s, workspaceName=%s]',
-            $this->contentStreamId->value,
+            'NodeAddress[dimensionSpacePoint=%s, nodeAggregateId=%s, workspaceName=%s]',
```
Comment (bwaidelich) [inline Neos.Neos/Classes/FrontendRouting/NodeAddress.php:56]: I hope that we don' relay on this string representation (=> do we really need it, what about using [__debugInfo()](https://www.php.net/manual/en/language.oop5.magic.php#object.debuginfo) instead?) – and if we keep it: Why did you decide to move the workspace name to the end?

File: Neos.Neos/Classes/Service/Controller/DataSourceController.php [hunk 6]
```diff
@@ -39,10 +47,9 @@ class DataSourceController extends AbstractServiceController
 
     /**
      * @param string $dataSourceIdentifier
-     * @param Node $node
      * @throws NeosException
      */
-    public function indexAction($dataSourceIdentifier, Node $node = null): void
+    public function indexAction($dataSourceIdentifier, string $node = null): void
```
Comment (bwaidelich) [inline Neos.Neos/Classes/Service/Controller/DataSourceController.php:52]: you removed the param annotation – I would instead suggest to keep it and add some docs instead – what is this `$node` string expected to contain?

File: Neos.Neos/Classes/Service/Controller/DataSourceController.php [hunk 7]
```diff
@@ -63,11 +70,29 @@ public function indexAction($dataSourceIdentifier, Node $node = null): void
...
 
-        $values = $dataSource->getData($node, $arguments);
+        $values = $dataSource->getData($this->deserializeNodeFromLegacyAddress($node), $arguments);
 
         $this->view->assign('value', $values);
     }
 
+    private function deserializeNodeFromLegacyAddress(?string $stringFormattedNodeAddress): ?Node
+    {
+        if (!$stringFormattedNodeAddress) {
```
Comment (bwaidelich) [inline Neos.Neos/Classes/Service/Controller/DataSourceController.php:80]:
```suggestion
        if ($stringFormattedNodeAddress === null) {
```

**Review by bwaidelich — COMMENTED** [review]
I just left some comments on first flyover, but I don't understand all the implications yet

File: Neos.Neos/Classes/Controller/Backend/ContentController.php (hunk 1 above)
Comment (mhsdesign) [reply Neos.Neos/Classes/Controller/Backend/ContentController.php:153]: jip a followup will change the whole neos ui and neos.neos to work on the new format we agreed upon.

File: Neos.Neos/Classes/FrontendRouting/NodeAddress.php (hunk 2 above)
Comment (mhsdesign) [reply Neos.Neos/Classes/FrontendRouting/NodeAddress.php:34]: because its just `@deprecated` and will be removed soon.
The statements are not true anymore: `It is used in Neos Routing to build a URI to a node.`

File: Neos.Neos/Classes/FrontendRouting/NodeAddress.php (hunk 3 above)
Comment (mhsdesign) [reply Neos.Neos/Classes/FrontendRouting/NodeAddress.php:35]: i dindt want to remove the param yet as it would break the neos ui now (we construct it over there) this is just intermediate

File: Neos.Neos/Classes/FrontendRouting/NodeAddress.php (hunk 4 above)
Comment (mhsdesign) [reply Neos.Neos/Classes/FrontendRouting/NodeAddress.php:37]: its just a dummy stub, and `_` indicates its unused. the class will be obliterated into nothing.

File: Neos.Neos/Classes/FrontendRouting/NodeAddress.php (hunk 5 above)
Comment (mhsdesign) [reply Neos.Neos/Classes/FrontendRouting/NodeAddress.php:56]: we probably dont even need the `__toString` anymore at this time ... but i kept it ... it will all go away. Its legacy.

File: Neos.Neos/Classes/Service/Controller/DataSourceController.php (hunk 6 above)
Comment (mhsdesign) [reply Neos.Neos/Classes/Service/Controller/DataSourceController.php:52]: as per my main description:

> where previously Node $node was used to trigger property mapping (DataSourceController and WorkspaceController) we now pass the node as string and use the legacy node address factory to handle the legacy format manually. A followup will adjust everything to work with the new node address.

**Review by mhsdesign — COMMENTED** [review]
regarding the _why_ of this change, i tried to put it into words in my main description above but i would also be happy to elaborate on that or have a quick huddle.

**Review by kitsunet — APPROVED** [review]
There is one little code suggestion from bwaidelich I would think should be merged, the rest seems fine to me.

**Comment by mhsdesign** [comment]
Okay in that case id merge it ;)

File: Neos.Neos/Classes/FrontendRouting/NodeAddress.php (hunk 2 above)
Comment (bwaidelich) [reply Neos.Neos/Classes/FrontendRouting/NodeAddress.php:34]: Right, thanks for clarifying, I briefly confused the two `NodeAddress`es.. Good that we are getting rid of one :)

File: Neos.Neos/Classes/Service/Controller/DataSourceController.php (hunk 6 above)
Comment (bwaidelich) [reply Neos.Neos/Classes/Service/Controller/DataSourceController.php:52]: OK, sorry, missed that one!
